### PR TITLE
Add OriginContext to track data across modules

### PR DESCRIPTION
**Pull request by allada**
Adds OriginContext that can have data set on it about the origin
and read out in any module at any time. This is designed to allow
us to track "requests" (usually network requests) as it traverses
through the system.

The most immediate use is to track what hashing algorithm was
requested for VerifyStore and verify the data using it.

In the future we'll likely use this for other things, like tracking
a "user" and emit what resources that user used over time and emit
it to another system based on a config.

- - -
This change is [Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875)

### CODE REVIEW

**Review by adam-singer — CHANGES_REQUESTED** [review]
> Reviewed 9 of 41 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875)* status: 0 of 1 LGTMs obtained, and pending CI: Remote / large-ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), integration-tests (20.04), integration-tests (22.04), and 4 discussions need to be resolved

___
*[`BUILD.bazel` line 6 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/875#-NwGFGdPC61uAUANCuYW:-NwGFGdPC61uAUANCuYX:bvovsio) ([raw file](https://github.com/TraceMachina/nativelink/blob/7ba6656ebbaa8d0a493d90a9f83fd7263dabc90a/BUILD.bazel#L6)):*
> ```Starlark
>     [
>         ".rustfmt.toml",
>         "clippy.toml",
> ```

nit `nativelink-clippy`? 
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1563 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/875#-NwGJ1xZ2V2iUt1zoZs7:-NwGJ1xZ2V2iUt1zoZs8:b-sadqpj) ([raw file](https://github.com/TraceMachina/nativelink/blob/7ba6656ebbaa8d0a493d90a9f83fd7263dabc90a/nativelink-scheduler/tests/simple_scheduler_test.rs#L1563)):*
> ```XML
>     #[tokio::test]
>     async fn ensure_task_or_worker_change_notification_received_test() -> Result<(), Error> {
>         OriginContext::init_for_test();
> ```

fwiw, we can probably make this a rust procedural macro out of this. Might also come in handy for other testing situations later when we are doing similar test setups verbosely over time

cargo deps

```
[lib]
proc-macro = true

[dependencies]
syn = { version = "1.0", features = ["full"] }
quote = "1.0"
```

example macro 

```
extern crate proc_macro;

use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, ItemFn};

#[proc_macro_attribute]
pub fn test_origin_context(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let input_fn = parse_macro_input!(item as ItemFn);

    let fn_name = &input_fn.sig.ident;
    let fn_block = &input_fn.block;
    let fn_inputs = &input_fn.sig.inputs;
    let fn_output = &input_fn.sig.output;

    let expanded = quote! {
        fn #fn_name(#fn_inputs) #fn_output {
            OriginContext::init_for_test();
            #fn_block
        }
    };

    TokenStream::from(expanded)
}
```

usage 

```
#[tokio::test]
#[test_origin_context]
    async fn ensure_task_or_worker_change_notification_received_test() -> Result<(), Error> {
```
___
*[`src/bin/nativelink.rs` line 525 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/875#-NwHN-Cm5NjFohtNzrO_:-NwHN-Cm5NjFohtNzrOa:bs3iz5w) ([raw file](https://github.com/TraceMachina/nativelink/blob/7ba6656ebbaa8d0a493d90a9f83fd7263dabc90a/src/bin/nativelink.rs#L525)):*
> ```XML
>                                 },
>                                 error_span!("set_drain_worker"),
>                                 Arc::new(OriginContext::default()),
> ```

Should each of the `default()` context calls be named?

**Review by allada — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875)* status: 0 of 1 LGTMs obtained, and pending CI: Analyze (javascript-typescript), Bazel Dev / ubuntu-22.04, Local / ubuntu-22.04, Vercel, pre-commit-checks, ubuntu-20.04 / stable, ubuntu-22.04, vale, and 2 discussions need to be resolved (waiting on @adam-singer)

___
*[`src/bin/nativelink.rs` line 525 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/875#-NwHN-Cm5NjFohtNzrO_:-NwSvRHL76FqtaCDjBCF:bnrkhzh) ([raw file](https://github.com/TraceMachina/nativelink/blob/7ba6656ebbaa8d0a493d90a9f83fd7263dabc90a/src/bin/nativelink.rs#L525)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

Should each of the `default()` context calls be named?
</blockquote></details>

Done. Used Builder pattern.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1563 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/875#-NwGJ1xZ2V2iUt1zoZs7:-NwSvZha6qslQm0Wc0L2:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/7ba6656ebbaa8d0a493d90a9f83fd7263dabc90a/nativelink-scheduler/tests/simple_scheduler_test.rs#L1563)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

fwiw, we can probably make this a rust procedural macro out of this. Might also come in handy for other testing situations later when we are doing similar test setups verbosely over time

cargo deps

```
[lib]
proc-macro = true

[dependencies]
syn = { version = "1.0", features = ["full"] }
quote = "1.0"
```

example macro 

```
extern crate proc_macro;

use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, ItemFn};

#[proc_macro_attribute]
pub fn test_origin_context(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let input_fn = parse_macro_input!(item as ItemFn);

    let fn_name = &input_fn.sig.ident;
    let fn_block = &input_fn.block;
    let fn_inputs = &input_fn.sig.inputs;
    let fn_output = &input_fn.sig.output;

    let expanded = quote! {
        fn #fn_name(#fn_inputs) #fn_output {
            OriginContext::init_for_test();
            #fn_block
        }
    };

    TokenStream::from(expanded)
}
```

usage 

```
#[tokio::test]
#[test_origin_context]
    async fn ensure_task_or_worker_change_notification_received_test() -> Result<(), Error> {
```
</blockquote></details>

Done.

**Review by allada — COMMENTED** [review]
This is ready for review.

> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875#-:-NwT5Yob1iTwLm4Slf_t:btjnoax)* status: 0 of 1 LGTMs obtained, and pending CI: Analyze (javascript-typescript), Analyze (python), Bazel Dev / ubuntu-22.04, Cargo Dev / macos-13, Cargo Dev / ubuntu-22.04, Local / ubuntu-22.04, Publish image, Publish nativelink-worker-lre-cc, Remote / large-ubuntu-22.04, asan / ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), integration-tests (20.04), integration-tests (22.04), macos-13, pre-commit-checks, ubuntu-20.04 / stable, ubuntu-22.04, ubuntu-22.04 / stable, windows-2022 / stable, and 2 discussions need to be resolved (waiting on @adam-singer)

**Review by adam-singer — APPROVED** [review]
There are no tests for this? 

> Reviewed 43 of 43 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875#-:-Nwg2PsH8LS1YaI6_ReD:b3ifrlv)* status: 0 of 1 LGTMs obtained, and pending CI: Remote / large-ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), integration-tests (20.04), integration-tests (22.04), and 4 discussions need to be resolved

___
*[`nativelink-util/src/origin_context.rs` line 29 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/875#-Nweg_DECJC7wRHwtjOw:-Nweg_DECJC7wRHwtjOx:b-96p0ln) ([raw file](https://github.com/TraceMachina/nativelink/blob/674f18f815fb1012d9370a79efcc2932aedf4edf/nativelink-util/src/origin_context.rs#L29)):*
> ```XML
> #[derive(Default)]
> pub struct OriginContext {
>     // TODO(allada) Add fields that need to be shared globally.
> ```

What are the thoughts on how fields are part of a context? In some frameworks usually this pattern is a Map of `key -> object` and helper methods to pull out the object as the right type. Given rust is a little bit different, curious what our patterns should be around this
___
*[`nativelink-util/src/origin_context.rs` line 32 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/875#-Nwg-cZRC8pkbMCfXY_C:-Nwg-cZRC8pkbMCfXY_D:b-l6go43) ([raw file](https://github.com/TraceMachina/nativelink/blob/674f18f815fb1012d9370a79efcc2932aedf4edf/nativelink-util/src/origin_context.rs#L32)):*
> ```XML
> }
> 
> impl OriginContext {
> ```

nit: Thoughts on why this should be called `OriginContext` vs `Context` and keeping the package/module as `origin_context`. I could see a reason to keep as is because `Context` might be overly general.

**Review by adam-singer — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875)* status: 0 of 1 LGTMs obtained, and pending CI: Remote / large-ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), integration-tests (20.04), integration-tests (22.04), and 2 discussions need to be resolved (waiting on @adam-singer)

**Review by adam-singer — APPROVED** [review]
> Reviewed 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875)* status: 0 of 1 LGTMs obtained, and pending CI: Remote / large-ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), integration-tests (20.04), integration-tests (22.04), and 2 discussions need to be resolved (waiting on @adam-singer)

**Review by allada — COMMENTED** [review]
I thought about writing a test or two, but there's nothing to really test yet. The next PR will be adding actual consumers to the `OrigionContext`, I think this is where we should be adding the tests. For example:
* Test to ensure verify store uses `OrigionContext`'s hash function when set.
* Test to ensure empty-blob hash returns proper hash if set in `OrigionContext`.
* (maybe-later) Test to ensure user is attributed to using execution units when set.

> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875#-:-Nwow0wg5MueahOAnq2M:bflzb)* status: 0 of 1 LGTMs obtained, and pending CI: Remote / large-ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), integration-tests (20.04), integration-tests (22.04), and 2 discussions need to be resolved (waiting on @adam-singer)

___
*[`nativelink-util/src/origin_context.rs` line 29 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/875#-Nweg_DECJC7wRHwtjOw:-Nwov_YM6LFaROj2PPqE:b-9xu59u) ([raw file](https://github.com/TraceMachina/nativelink/blob/674f18f815fb1012d9370a79efcc2932aedf4edf/nativelink-util/src/origin_context.rs#L29)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

What are the thoughts on how fields are part of a context? In some frameworks usually this pattern is a Map of `key -> object` and helper methods to pull out the object as the right type. Given rust is a little bit different, curious what our patterns should be around this
</blockquote></details>

I was going back and forth on what pattern we'll use. I have settled on one of two options, but it's not introduced in this PR, so lets wait till next PR to discuss how we transport the actual data.
___
*[`nativelink-util/src/origin_context.rs` line 32 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/875#-Nwg-cZRC8pkbMCfXY_C:-NwovmX90fWqL5UXDKu4:bu3hfqb) ([raw file](https://github.com/TraceMachina/nativelink/blob/674f18f815fb1012d9370a79efcc2932aedf4edf/nativelink-util/src/origin_context.rs#L32)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

nit: Thoughts on why this should be called `OriginContext` vs `Context` and keeping the package/module as `origin_context`. I could see a reason to keep as is because `Context` might be overly general.
</blockquote></details>

Yeah, I thought about `Context`, but because rust's futures use `Context` and other packages also use context, I didn't want to overload the name, so settled on `OriginContext`.

**Review by adam-singer — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875#-:-NwzWo_oBDk-1bIl0ARO:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 1 LGTMs obtained

**Review by allada — COMMENTED** [review]
After starting to implement this for `DigestFunction`, I realized I didn't like the API, so I rewrote it to be more natural to use.

> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875#-:-NxAUhzu9XTcyLJvVoOX:b-bgelb8)* status: 1 of 1 LGTMs obtained, and pending CI: Bazel Dev / ubuntu-22.04, Publish image, Vercel, pre-commit-checks, ubuntu-22.04, vale

File: nativelink-util/src/origin_context.rs [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+/// Make a symbol that represents a unique memory pointer that is
+/// constant and choosen at compile time. This enables two modules
+/// to use that memory location to reference data that lives in a
+/// shared module without the two modules knowing about eachother.
+/// For example, lets say we have a context that holds anonymous
+/// data; we can use these symbols to let one module set the data
+/// and tie the data to a symbol and another module read the data
+/// with the symbol, without the two modules knowing about eachother.
+#[macro_export]
+macro_rules! make_symbol {
```
Comment (MarcusSorealheis) [inline nativelink-util/src/origin_context.rs:41]: :)

**Review by MarcusSorealheis — APPROVED** [review]

**Review by allada — COMMENTED** [review]
> Dismissed @MarcusSorealheis from a discussion.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/875)* status: <span>:shipit:</span> complete! 1 of 1 LGTMs obtained

___
*[`nativelink-util/src/origin_context.rs` line 41 at r5](https://reviewable.io/reviews/TraceMachina/nativelink/875#gh-1591690637:-NxJBKO5DZ9UBCQQPvkW:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/c63c7ad41b1dcd2be3e54ed418f69c537985c619/nativelink-util/src/origin_context.rs#L41)):*
<details><summary><i>Previously, MarcusSorealheis (Marcus Eagan) wrote…</i></summary><blockquote>

:) 
</blockquote></details>

Done.

**Pull request merged**